src/generate.ts

```diff
@@ -22,5 +28,101 @@ export function generate(specQ: SpecQuery, schema: Schema, opt: QueryConfig = DE
     }
   });
 
+  let ENCODING_QUERY_INDEX: Dict<EncodingQuery> = {};
+
+  if (opt.smallBandSizeForHighCardinality || opt.smallBandSizeForFacet) {
+    answerSet = answerSet.map(function(specQM) {
+
+      [Channel.ROW, Channel.Y, Channel.COLUMN, Channel.X].forEach((channel) => {
+        ENCODING_QUERY_INDEX[channel] = specQM.getEncodingQueryByChannel(channel);
+      });
+
+      const yEncQ = ENCODING_QUERY_INDEX[Channel.Y];
+      if (yEncQ !== undefined) {
+        if (ENCODING_QUERY_INDEX[Channel.ROW] ||
+            schema.cardinality(yEncQ) > 10) {
+
+          if (yEncQ.scale === undefined) {
+            specQM.setEncodingProperty(
+              specQM.getEncodingQueryIndexByChannel(Channel.Y),
+              Property.SCALE,
+              true,
+              {name: 'scale', values: [true, false]} // not sure what this EnumSpec should be
+            );
+          }
+
+          if (yEncQ.scale) {
+            specQM.setEncodingProperty(
+              specQM.getEncodingQueryIndexByChannel(Channel.Y),
+              Property.SCALE_BANDSIZE,
+              12,
+              {name: 'scaleBandSize', values: [12]} // not sure what this EnumSpec should be
+            );
+          }
+        }
+      }
+
+      const xEncQ = ENCODING_QUERY_INDEX[Channel.X];
+      if (xEncQ !== undefined) {
+        if (ENCODING_QUERY_INDEX[Channel.COLUMN] ||
+            schema.cardinality(xEncQ) > 10) {
+
+          if (xEncQ.scale === undefined) {
+            specQM.setEncodingProperty(
+              specQM.getEncodingQueryIndexByChannel(Channel.X),
+              Property.SCALE,
+              true,
+              {name: 'scale', values: [true, false]} // not sure what this EnumSpec should be
+            );
+          }
+
+          if (xEncQ.scale) {
+            specQM.setEncodingProperty(
+              specQM.getEncodingQueryIndexByChannel(Channel.X),
+              Property.SCALE_BANDSIZE,
+              12,
+              {name: 'scaleBandSize', values: [12]} // not sure what this EnumSpec should be
+            );
+          }
+        }
+      }
+
+      return specQM;
+    });
+  }
+
+  if (opt.nominalScaleForHighCardinality) {
+    answerSet = answerSet.map(function(specQM) {
```

Comment:
Write this in a way that call map only once; otherwise, it's inefficient to create new array multiple times.  Inside map you consider multiple modifiers (call multiple modifier methods).